app/routines/create_user.rb

```diff
@@ -41,19 +29,44 @@ def exec(username:, title: nil, first_name: nil, last_name: nil,
     transfer_errors_from(outputs[:user], {type: :verbatim})
   end
 
-  # Returns a valid, though possibly non-unique, username
-  def get_valid_username(base)
-    base = randomify_username(base) if base.blank?
+  def generate_unique_valid_username(username)
+    return username if username_is_valid?(username) && username_is_unique?(username)
+
+    loop do
+      username = create_random_username
+      break if username_is_unique?(username) || max_attempts_exceeded?
+    end
+
+    username
+  end
+
+  def max_attempts_exceeded?
+    @attempts_count ||= 0
+    @attempts_count += 1
+    did_exceed = @attempts_count > username_max_attempts
+    did_exceed
+  end
```

Comment:
@BryanEliDimas I agree with @jpslav: `max_attempts_exceeded?` should probably be inlined back into `generate_unique_valid_username`.  Also, if the number of attempts exceeds the max and `ensure_no_errors` is set, it seems like we should raise an exception instead of just breaking the loop since that is the contract this routine implies.